File: LeetCode/Liang/2AddTwoNumbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    // Be careful [5] [5]
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *dummy = new ListNode(INT_MIN);
        ListNode *tail = dummy;
        int carry = 0;
        while (l1 || l2) {
            int num1 = 0, num2 = 0;
            if (l1) {
                num1 = l1->val;
                l1 = l1->next;
            }
            
            if (l2) {
                num2 = l2->val;
                l2 = l2->next;
            }
            
            int sum = num1 + num2 + carry;
            carry = sum / 10;
            ListNode *newNode = new ListNode(sum % 10);
            tail->next = newNode;
            tail = tail->next;
        }
        
        if (carry != 0) {
            ListNode *newNode = new ListNode(carry);
            tail->next = newNode;
            tail = tail->next;
        }
        
        ListNode *head = dummy->next;
        delete dummy;
        return head;
    }
};

File: LeetCode/Hao/2AddTwoNumbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode result = new ListNode(0);
        ListNode p = result;
        int carry = 0;
        
        while (l1 != null || l2 != null) {
            int number = 0;
            if (l1 == null) {
                number = l2.val + carry;
                l2 = l2.next;
            } else if (l2 == null) {
                number = l1.val + carry;
                l1 = l1.next;
            } else {
                number = l1.val + l2.val + carry;
                l1 = l1.next;
                l2 = l2.next;
            }
            
            carry = number >= 10 ? 1 : 0;
            number %= 10;
            
            result.next = new ListNode(number);
            result = result.next;
        }
        
        if(carry != 0) {
            result.next = new ListNode(1);
        }
        
        
        return p.next;
        
    }
}
